### ISSUE TITLE
DASH Low Latency on ExoPlayer version 2.13.2 issue frame freeze

### ISSUE THREAD

**Issue by hot1sp** [google/ExoPlayer#9092]
Hi

We're testing the DASH Low Latency on ExoPlayer version 2.13.2 and encountering an issue of freeze frame during playback

The issue is being described as a freeze frame happens from time to time after tune to the LLC channel, which we can see in the player logs accordingly a Buffering message (freeze occurs) and after 1-2 seconds resume to Ready and continues playback:

```
2021-06-16 13:25:34.428 8665-8665/E/Test Test ExoPlayPlaybackState: Buffering
2021-06-16 13:25:34.448 8665-9325/D/BufferPoolAccessor: bufferpool2 0xb74f2310 : 4(32768 size) total buffers - 4(32768 size) used buffers - 0/4 (recycle/alloc) - 4/753 (fetch/transfer)
2021-06-16 13:25:35.393 8665-8665/E/Test Test ExoPlayPlaybackState: Ready
``` 

The DASH LLC stream under test contained fragment length of 2 seconds (where IDR interval of the stream is also 2 seconds), with 2 layers of bit rate, and frame rate of 50fps (please find attached DASH MPD file): [Manifest_DASH_LLC_FL_2_Sec.txt](https://github.com/google/ExoPlayer/files/6686834/Manifest_DASH_LLC_FL_2_Sec.txt)

Knowing that with such length of fragment the buffer on player could be decreased, We had tried to adjust the following buffering parameters on ExoPlayer in order to avoid that freeze frames but it didn't resolve our issue:

bufferForPlaybackMs (decreased from 2500 to 1000)
minBufferMs (decreased from 15000 to 1000)
maxBufferMs (decreased from 15000 to 1000)
bufferForPlaybackAfterRebufferMs (decreased from 5000 to 1000)

please advise how should we proceed further to have the playback without freeze frames 

thanks.

**Comment by icbaker** [comment]
I'm not completely clear of the problem you're describing. Do you see playback of the attached manifest start, progress and then pause briefly before continuing? Or are you asking about start-up latency - i.e. the time before the content starts playing at all? It's not obvious what you mean by 'tune to the LLC channel', since of course there's no 'tuning' into an RF signal going on :)

The provided manifest is not enough for us to play the content to try and reproduce - the content is referenced with relative URLs, so we need a link to the manifest that's hosted alongside the content. We probably also need the DRM license server details. Please either post this info here or send to dev.exoplayer@gmail.com using a subject in the format "Issue #1234" (where "#1234" should be replaced with this issue number.) Please also update this issue to indicate you’ve done this.

**Comment by hot1sp** [comment]
Hi

 

We're talking about zapping to low latency live OTT DASH channel (which we have sent to you its manifest file) – do you see an issue with this manifest ?

After zapping to that OTT channel, it plays but is suffering from frequent freeze frames and then continue to play, and then again video frame freeze…and so on (during that freeze we can see buffering alerts on client log)

Can you advise on which parameters to adjust on ExoPlayer to improve the playback ?

 

We can provide the content stream URL to the manifest of that channel and the URL to DRM License, but we need to open for you a specific access to it

Can you please share with yours public source specific IP address from which you can test that channel with yours ExoPlayer ?

 

thanks

**Comment by icbaker** [comment]
> After zapping to that OTT channel, it plays but is suffering from frequent freeze frames and then continue to play, and then again video frame freeze…and so on (during that freeze we can see buffering alerts on client log)
> 
> Can you advise on which parameters to adjust on ExoPlayer to improve the playback ?

Are the freezes you see aligned with segment boundaries, or do they happen within segments?

From the information provided it sounds like the network connection between the device & the content server isn't fast enough to deliver the content without buffering? If that's the case then there's no changes you can make to ExoPlayer which will resolve the problem, you either need to speed up the network connection or reduce the bitrate of the content. 

Or maybe the server isn't serving content far enough in advance of the 2s - 8s live edge defined by the `<Latency>` node in the manifest?

> bufferForPlaybackMs (decreased from 2500 to 1000)
> minBufferMs (decreased from 15000 to 1000)
> maxBufferMs (decreased from 15000 to 1000)
> bufferForPlaybackAfterRebufferMs (decreased from 5000 to 1000)

If the network is jittery (i.e. throughput & latency keeps varying) then these load control changes you made will **increase** the chance of repeated rebuffers (because you're asking the player to start playback with less content buffered, meaning it's less robust against some sudden slowness). For an extreme (and obviously unrealistic) example: If you configured these values to be longer than your content, you'd guarantee your viewers would never see a rebuffer (because all the content is downloaded before playback starts) - in exchance for significantly increased latency before playback starts.

It's possible I'm misunderstanding the problem though. A bug report (`adb bugreport`) taken after reproducing the problem in the demo app might contain more helpful information. Please either upload it here or send to dev.exoplayer@gmail.com using a subject in the format "Issue #1234" (where "#1234" should be replaced with this issue number.) Please also update this issue to indicate you’ve done this.

> We can provide the content stream URL to the manifest of that channel and the URL to DRM License, but we need to open for you a specific access to it
> 
> Can you please share with yours public source specific IP address from which you can test that channel with yours ExoPlayer ?

I'm afraid I can't provide that information - the provided media needs to be globally accessible for me to try and reproduce the problem.

**Comment by hot1sp** [comment]
Hi


Indeed we can see a significant improvement on the DASH Live low latency playback performance when connecting the client with the ExoPlayer into a high speed Internet link

Previously were connected with up to 100Mbps link on the Client, and issue of freeze frame observed – now with above 100Mbps link (potentially 200Mbps – 300Mbps) the playback is good, no freeze frame observed

1. Do you familiar with such a pre-requisite on the Network link between the DASH low latency streaming server to the client device that should be more than 100Mbps for playback to have high performance without freeze frames ?

2. Which parameter on the DASH manifest of low latency channel tells the ExoPlayer the amount of short buffering/ how much fast it should release the chunks from the buffer to actual playback ?

Another issue occurred on that matter is that the manifest contain 2 bit rate Levels (higher and lower ones), so during tests of ABR we loaded the network and the client moved from the high bit rate level (4.6Mps) into the lower one (2.5Mbps) and played it as expected – But when we released that load from the network the client still played the lower bit rate level (2.5Mbps) and ONLY after zapping again to that channel then the client played the higher bit rate level:

3. My question is if there a parameter on ExoPlayer to adjust on the ABR with low latency DASH channel so that player will know that when the stress on the network released and there's enough bandwidth so it can go back from the lower level to the higher bit rate level on the ABR ?

Our manifest of the low latency DASH Live channel contains the following Latency parameters:
Latency target="4000" min="2000" max="8006"

1.Can you please explain how the ExoPlayer is using each of those parameters ? Latency target, min and max

2.What's the meaning of each parameter ?

3.What's the recommendation on the values to put on those parameters in order to have good playback performance on the ExoPlayer 2.13.2 ?

4.On the DASH MPD of our Low Latency Live channel, the minBufferTime parameter is being set to 2 seconds

5.Could be that 2 seconds is too low for ExoPlayer to keep some buffer ?

6.Should we adjust this parameter into a higher value such as 3 seconds ?

Thanks

**Comment by icbaker** [comment]
> Indeed we can see a significant improvement on the DASH Live low latency playback performance when connecting the client with the ExoPlayer into a high speed Internet link
> 
> Previously were connected with up to 100Mbps link on the Client, and issue of freeze frame observed – now with above 100Mbps link (potentially 200Mbps – 300Mbps) the playback is good, no freeze frame observed
> 
> 1. Do you familiar with such a pre-requisite on the Network link between the DASH low latency streaming server to the client device that should be more than 100Mbps for playback to have high performance without freeze frames ?

Just to check: You've measured a throughput of 100Mbps on the 'slow' link? When the throughput is that much higher than the required capacity (~2.5Mbps in this case?) but you're still seeing problems it seems like other properties of the link must be causing you problems. e.g. is it burst-y, so fast for a bit then slow for a bit, or is the latency really jittery? If it averages 100Mbps because it goes at 1Gbps for 1s then 0Mbps for the next 9 seconds then you can see how there would be problems with video playback that's only buffering a few seconds.

> 2. Which parameter on the DASH manifest of low latency channel tells the ExoPlayer the amount of short buffering/ how much fast it should release the chunks from the buffer to actual playback ?

As you said below, your manifest says `Latency target="4000" min="2000" max="8006"`. That means ExoPlayer sets these values into its `targetOffsetMs`, `minOffsetMs` and `maxOffsetMs` fields. More info on how these are used: https://exoplayer.dev/live-streaming.html.

> Another issue occurred on that matter is that the manifest contain 2 bit rate Levels (higher and lower ones), so during tests of ABR we loaded the network and the client moved from the high bit rate level (4.6Mps) into the lower one (2.5Mbps) and played it as expected – But when we released that load from the network the client still played the lower bit rate level (2.5Mbps) and ONLY after zapping again to that channel then the client played the higher bit rate level:
> 
> 1. My question is if there a parameter on ExoPlayer to adjust on the ABR with low latency DASH channel so that player will know that when the stress on the network released and there's enough bandwidth so it can go back from the lower level to the higher bit rate level on the ABR ?

Measuring bandwidth while playing low latency live content is very hard, because it's impossible for the player to determine if the speed that it's downloading content is limited by the network link or by the server making content available. I suspect that might be why you don't see ABR switching back up when the network recovers. @christosts might have thoughts here.

> 
> Our manifest of the low latency DASH Live channel contains the following Latency parameters:
> Latency target="4000" min="2000" max="8006"
> 
> 1.Can you please explain how the ExoPlayer is using each of those parameters ? Latency target, min and max
> 
> 2.What's the meaning of each parameter ?

Answered above
> 
> 3.What's the recommendation on the values to put on those parameters in order to have good playback performance on the ExoPlayer 2.13.2 ?
> 
> 4.On the DASH MPD of our Low Latency Live channel, the minBufferTime parameter is being set to 2 seconds
> 
> 5.Could be that 2 seconds is too low for ExoPlayer to keep some buffer ?
> 
> 6.Should we adjust this parameter into a higher value such as 3 seconds ?

I'm afraid preparing media is outside the scope of this issue tracker. It's not something we have the expertise to help you with.

**Comment by hot1sp** [comment]
Thank you very much for the quick answers. 
I have few more questions:
1.To which parameter value the ExoPlayer set the following parameter coming in Manifest ? "MinBufferTime"
2. Can you please provide more information on the usage of this parameter on ExoPlayer ?
 I didn't realize it from the Documentation.
 Thanks.

**Comment by christosts** [comment]
> Measuring bandwidth while playing low latency live content is very hard, because it's impossible for the player to determine if the speed that it's downloading content is limited by the network link or by the server making content available. I suspect that might be why you don't see ABR switching back up when the network recovers. @christosts might have thoughts here.

In Low-Latency live streaming, the player cannot make bandwidth measurements when downloading segments close to the live edge because these segments are not downloaded at full speed, therefore they are not good sample points for estimating bandwidth throughput. As a result, when the player is streaming Low-Latency live streams:
- it uses any preexisting bandwidth estimation and does not change it while playing close to the live edge
- if the player rebuffers and has to pause, then it will increase the offset (it will stay back a bit) and download chunks that are slightly older. Such chunks are likely to be transferred at full speed, hence give the player a new measurements for bandwidth estimate.

That said, performing ABR with Low-Latency is hard and there is definitely space for improvements. For example, high latency or jitter are not currently taken into account and these may affect Low-Latency significantly. You mentioned that your network is at 100Mbps, but it could be suffering from high latency to the server.

> 1.To which parameter value the ExoPlayer set the following parameter coming in Manifest ? "MinBufferTime"

The parameter is stored in [DashManifest](https://github.com/google/ExoPlayer/blob/r2.14.0/library/dash/src/main/java/com/google/android/exoplayer2/source/dash/manifest/DashManifest.java) instances and it is used internally in ExoPlayer. I don't think there's a configurable parameter that maps *directly* to this parameter (please note the use of *directly* - more below).

> 2. Can you please provide more information on the usage of this parameter on ExoPlayer ?

For the case of Low-Latency live streaming, the DASH manifest's `MinBufferTime` is used to define the player's `minOffsetMs` **if** the user has not set their own `LiveConfiguration` parameters in MediaItem, or the manifest does not contain `Latency`. You have said previously that your manifest defines Latency already, in which case ExoPlayer will adjust its targetOffset/minOffset/maxOffset from there. May I suggest you increase both targetOffset and minOffset to ensure the player stays a bit behind the live edge? This will come at the cost of increased latency, but it might remedy the rebuffers.

Also, in addition to the [documentation page](https://exoplayer.dev/live-streaming.html) provided by @icbaker, you may find more information in our [blog post](https://medium.com/google-exoplayer/low-latency-live-streaming-with-exoplayer-8552d5841060) on Low-Latency.

**Comment by hot1sp** [comment]
Hi


We had noticed that once launching the Exoplayer with the DASH low latency URL, so in the first time that Exoplayer plays this URL the lower level of ABR being played and the player is not moving into higher ABR level.

Only when zapping again to that channel and calling the same stream URL in the 2nd time, then the higher ABR level played 

In which parameter we can set the player to play initially (in the first time calling stream URL) the higher ABR level of the DASH Low latency channel ?

thx

**Comment by christosts** [comment]
I'd suggest to try tuning the [DefaultBandwidthMeter](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/upstream/DefaultBandwidthMeter.html).

The player uses a [BandwidthMeter](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/upstream/BandwidthMeter.html), to measure bandwidth throughput for ABR. By default, the player uses an instance of `DefaultBandwidthMeter`. Initially, while there have been no downloads, the `DefaultBandwidthMeter` is feeding the ABR algorithm with pre-determined bandwidth estimates based on location (country) and network type (wifi/mobile). If your first playback attempt is a low-latency stream, it's likely the `DefaultBandwidthMeter` will not be able to get new throughput measurements and keep using the pre-determined bandwidth estimate. It's also possible that the predetermined bandwidth estimate does not reflect the real conditions of your tests. Please try to create a custom DefaultBandwidthMeter with an initial bitrate that better matches your real conditions and pass it to the player during construction, for example

```
SimpleExoPlayer player =
  new SimpleExoPlayer.Builder(context)
     .setBandwidthMeter(new DefaultBandwidthMeter.Builder(context).setInitialBitrateEstimate(...).build())
     .build();
```

**Comment by hot1sp** [comment]
Thank you ,it helped. 
Now I can see in both WIFI and Ethernet the high quality profile selected full HD at start.
